exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,624 @@
+        } else {
+            Ok(())
+        }
+    }
+
+    fn add_deployed_artifact(&mut self, id: JavaArtifactId) {
+        self.deployed_artifacts.insert(id);
+    }
+
+    fn can_start_service(&self, instance: &Instance) -> Result<(), ExecutionError> {
+        if self.started_services.contains_key(&instance.id) {
+            Err(Error::AlreadyStarted.into())
+        } else {
+            Ok(())
+        }
+    }
+
+    fn add_started_service(&mut self, instance: Instance) {
+        self.started_services_by_name
+            .insert(instance.name.clone(), instance.id);
+        self.started_services.insert(instance.id, instance);
+    }
+
+    fn remove_started_service(&mut self, id: &InstanceId) {
+        let service = self.started_services.remove(id);
+        if let Some(instance) = service {
+            self.started_services_by_name.remove(&instance.name);
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
+        res.map_err(|err| match err.0 {
+            JniErrorKind::JavaException => Error::JavaException.into(),
+            _ => Error::UnspecifiedError.into(),
+        })
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(
+            ErrorKind::runtime(value as u8),
+            format!("{:?}", value.clone()),
+        )
+    }
+}
+
+impl Runtime for JavaRuntimeProxy {
+    fn deploy_artifact(
+        &mut self,
+        artifact: ArtifactId,
+        deploy_spec: Any,
+    ) -> Box<dyn Future<Item = (), Error = ExecutionError>> {
+
+        let id = match self.parse_artifact(&artifact) {
+            Ok(id) => id,
+            Err(err) => return Box::new(Err(err).into_future()),
+        };
+
+        if let Err(err) = self.can_deploy_artifact(&id) {
+            return Box::new(Err(err).into_future());
+        }
+
+        Box::new(Self::parse_jni(self.exec.with_attached(|env| {
+            let artifact_id = JObject::from(env.new_string(id.to_string())?);
+            let spec = JObject::from(env.byte_array_from_slice(&deploy_spec.into_bytes())?);
+
+            env.call_method_unchecked(
+                self.runtime_adapter.as_obj(),
+                runtime_adapter::deploy_artifact_id(),
+                JavaType::Primitive(Primitive::Void),
+                &[
+                    JValue::from(artifact_id),
+                    JValue::from(spec),
+                ],
+            )?;
+            Ok(())
+        })).map(|result| {
+            self.add_deployed_artifact(id);
+            result
+        })
+        .into_future())
+    }
+
+    fn artifact_protobuf_spec(&self, id: &ArtifactId) -> Option<ArtifactProtobufSpec> {
+        let id = self.parse_artifact(id).ok()?;
+
+        if self.deployed_artifacts.contains(&id) {
```

Comment:
Why this check is needed? I'd assume that the framework won't call the runtimes
that don't have a given artifact. The runtime implementation (in Java) might get this check if it gets implemented, but it does not seem to be needed in the proxy.